### PR TITLE
[PD$-110002] Part 9: Cassandra Client Pool II - Distribution Outlier Controllers

### PR DESCRIPTION
**Pull request by jeremyk-91**
**WIP**: Reviewable, but needs tests.

**Goals (and why)**:
- $$$
- With many Cassandra metrics, we aren't particularly interested in the individual pool values. We are interested in the *mean* of the values, as well as if there are any *outliers*. This PR attempts to capture that sentiment: this could be useful, because at large stacks we could save 5 * |namespaces| * (|cassandra cluster| - 1) time series in atlasdb-proxy, and 5 * (|cassandra cluster| - 1) time series elsewhere.

**Implementation Description (bullets)**:
- Implement a `DistributionOutlierController` that tracks a mean-gauge that is re-computed every 30 seconds, and creates publication filters for the constituent metrics that publish if they are more than the mean multiplied by a constant, or less than the mean multiplied by a less-than-1 constant.

**Testing (What was existing testing like?  What have you done to improve it?)**:
On its way: I'd add unit tests + do a bit of manual verification.

**Concerns (what feedback would you like?)**:
- If all nodes are wildly divergent and/or there are extremely bimodal distributions we might actually end up producing more metrics, though looking at internal metrics portal this doesn't seem to be the case.
- Implementing a filter to not publish the mean if everything was individually published seemed messy.
- Is `2.0` not aggressive enough (or too aggressive) as a constant?
- Are things wired correctly?

**Where should we start reviewing?**: DistributionOutlierController

**Priority (whenever / two weeks / yesterday)**: this week

### CODE REVIEW

File: atlasdb-cassandra/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/pool/CassandraClientPoolMetricsTest.java [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+                .containsKey(createMeanActiveTimeMillisMetric("mean"));
+
+        assertThat(metricsManager.getPublishableMetrics().getMetrics())
+                .doesNotContainKey(createMeanActiveTimeMillisMetric("pool1"))
+                .doesNotContainKey(createMeanActiveTimeMillisMetric("pool2"))
+                .containsKey(createMeanActiveTimeMillisMetric("pool3"))
+                .containsKey(createMeanActiveTimeMillisMetric("mean"));
+    }
+
+    public MetricName createMeanActiveTimeMillisMetric(String pool1) {
```
Comment (jeremyk-91) [inline atlasdb-cassandra/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/pool/CassandraClientPoolMetricsTest.java:64]: wat

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/pool/DistributionOutlierController.java [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+    private final double minimumMeanMultiple;
+    private final double maximumMeanMultiple;
+    private final Gauge<Double> meanGauge;
+
+    @VisibleForTesting
+    DistributionOutlierController(
+            Clock clock,
+            double minimumMeanMultiple,
+            double maximumMeanMultiple) {
+        this.gauges = new HashSet<>();
```
Comment (jeremyk-91) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/pool/DistributionOutlierController.java:52]: ConcurrentHashSet

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/pool/CassandraClientPoolMetrics.java [hunk 3]
```diff
@@ -72,6 +102,33 @@ public void recordPoolExhaustion() {
...
+            int poolNumber,
+            MetricPublicationFilter filter) {
+        Map<String, String> poolTag = ImmutableMap.of("pool", "pool" + poolNumber);
+        metricsManager.addMetricFilter(
+                CassandraClientPoolingContainer.class,
+                metric.metricName,
+                poolTag,
+                filter);
+        metricsManager.registerOrGet(
+                CassandraClientPoolingContainer.class,
```
Comment (sudiksha27) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/pool/CassandraClientPoolMetrics.java:126]: 💯

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/pool/DistributionOutlierController.java [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+    }
+
+    Gauge<Double> getMeanGauge() {
+        return meanGauge;
+    }
+
+    MetricPublicationFilter registerAndCreateFilter(Gauge<Long> gauge) {
+        gauges.add(gauge);
+        return () -> shouldPublishIndividualGaugeMetric(gauge);
+    }
```
Comment (sudiksha27) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/pool/DistributionOutlierController.java:84]: nice

**Pull request merged**